buildSrc/src/main/java/org/elasticsearch/gradle/precommit/FilePermissionsTask.java

```diff
@@ -0,0 +1,112 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.gradle.precommit;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.attribute.PosixFileAttributeView;
+import java.nio.file.attribute.PosixFilePermission;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.apache.tools.ant.taskdefs.condition.Os;
+import org.gradle.api.DefaultTask;
+import org.gradle.api.GradleException;
+import org.gradle.api.file.FileCollection;
+import org.gradle.api.plugins.JavaPluginConvention;
+import org.gradle.api.tasks.InputFiles;
+import org.gradle.api.tasks.OutputFile;
+import org.gradle.api.tasks.SourceSetContainer;
+import org.gradle.api.tasks.StopExecutionException;
+import org.gradle.api.tasks.TaskAction;
+import org.gradle.api.tasks.util.PatternFilterable;
+import org.gradle.api.tasks.util.PatternSet;
+
+/**
+ * Checks source files for correct file permissions.
+ */
+public class FilePermissionsTask extends DefaultTask {
+
+    /**
+     * A pattern set of which files should be checked.
+     */
+    private final PatternFilterable filesFilter = new PatternSet()
+            // we always include all source files, and exclude what should not be checked
+            .include("**")
+            // exclude sh files that might have the executable bit set
+            .exclude("**/*.sh");
+
+    private File outputMarker = new File(getProject().getBuildDir(), "markers/filePermissions");
+
+    public FilePermissionsTask() {
+        setDescription("Checks java source files for correct file permissions");
+    }
+
+    private static boolean isExecutableFile(File file) {
+        try {
+            Set<PosixFilePermission> permissions = Files.getFileAttributeView(file.toPath(), PosixFileAttributeView.class)
+                    .readAttributes()
+                    .permissions();
+            return permissions.contains(PosixFilePermission.OTHERS_EXECUTE)
+                    || permissions.contains(PosixFilePermission.OWNER_EXECUTE)
+                    || permissions.contains(PosixFilePermission.GROUP_EXECUTE);
+        } catch (IOException e) {
+            throw new IllegalStateException("unable to read the file " + file + " attributes", e);
+        }
+    }
+
+    /**
+     * Returns the files this task will check
+     */
+    @InputFiles
+    public FileCollection getFiles() {
+        SourceSetContainer sourceSets = getProject().getConvention().getPlugin(JavaPluginConvention.class).getSourceSets();
+        return getProject().files(sourceSets.stream()
+                .map(sourceSet -> sourceSet.getAllSource().matching(filesFilter))
+                .toArray());
```

Comment:
`FileTree` extends `FileCollection`, so neither `getProject().files(` nor `.toArray()` are necessary.

Reply:
sorry, I do not know the API that well...
I tried to fix this but I am not sure how to handle the case when no source matches? (I do not like the .orElse(null))

Reply:
The `.orElse(null)` worked in the unit tests but wouldn't have worked in an integration test as the method is annotated as input and thus can't return null. I took the liberty to push a fix for this.